kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -204,55 +230,119 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
+
+    val kmClass = rawType.toKmClass(throwOnNotClass = true) ?: return null
+
+    require(!Flag.IS_ABSTRACT(kmClass.flags)) {
       "Cannot serialize abstract class ${rawType.name}"
     }
-    require(!rawTypeKotlin.isInner) {
+    require(!Flag.Class.IS_INNER(kmClass.flags)) {
       "Cannot serialize inner class ${rawType.name}"
     }
-    require(rawTypeKotlin.objectInstance == null) {
+    require(!Flag.Class.IS_OBJECT(kmClass.flags)) {
       "Cannot serialize object declaration ${rawType.name}"
     }
-    require(!rawTypeKotlin.isSealed) {
+    require(!Flag.Class.IS_COMPANION_OBJECT(kmClass.flags)) {
+      "Cannot serialize companion object declaration ${rawType.name}"
+    }
+    require(!Flag.IS_SEALED(kmClass.flags)) {
       "Cannot reflectively serialize sealed class ${rawType.name}. Please register an adapter."
     }
 
-    val constructor = rawTypeKotlin.primaryConstructor ?: return null
-    val parametersByName = constructor.parameters.associateBy { it.name }
-    constructor.isAccessible = true
+    val kmConstructor = kmClass.constructors.find { Flag.Constructor.IS_PRIMARY(it.flags) }
+        ?: return null
+    val kmConstructorSignature = kmConstructor.signature?.asString() ?: return null
+    val constructorsBySignature = rawType.declaredConstructors.associateBy { it.jvmMethodSignature }
+    val jvmConstructor = constructorsBySignature[kmConstructorSignature] ?: return null
+    val parameterAnnotations = jvmConstructor.parameterAnnotations
+    val parameterTypes = jvmConstructor.parameterTypes
+    val parameters = kmConstructor.valueParameters.withIndex()
+        .map { (index, kmParam) ->
+          ParameterData(kmParam, index, parameterTypes[index], parameterAnnotations[index].toList())
+        }
+
+    val anyOptional = parameters.any { it.declaresDefaultValue }
+    val actualConstructor = if (anyOptional) {
+      val prefix = jvmConstructor.jvmMethodSignature.removeSuffix(")V")
+      val parameterCount = jvmConstructor.parameterTypes.size
+      val maskParamsToAdd = if (parameterCount == 0) {
+        0
+      } else {
+        (parameterCount + 31) / 32
+      }
+      val defaultConstructorSignature = buildString {
+        append(prefix)
+        repeat(maskParamsToAdd) {
+          append("I")
+        }
+        append(Util.DEFAULT_CONSTRUCTOR_MARKER!!.descriptor)
+        append(")V")
+      }
+      constructorsBySignature[defaultConstructorSignature] ?: return null
+    } else {
+      jvmConstructor
+    }
+
+    val constructorData = ConstructorData(
+        rawType,
+        kmConstructor,
+        actualConstructor,
+        parameters,
+        anyOptional
+    )
+    val parametersByName = parameters.associateBy { it.name }
 
     val bindingsByName = LinkedHashMap<String, KotlinJsonAdapter.Binding<Any, Any?>>()
 
-    for (property in rawTypeKotlin.memberProperties) {
-      val parameter = parametersByName[property.name]
+    // TODO this doesn't cover platform types
```

Comment:
what does this mean we have to do?

Reply:
Also from the PR description, we should just make a call here